# Data Analysis in Python

These datasets contain information about orders, customers, sellers, order items, and reviews, which will be used in further analysis.

In [1]:
import pandas as pd

## Extract
In this step, the required datasets are loaded using pandas.


In [2]:
customers = pd.read_csv('../data/olist_customers_dataset.csv')
order_items = pd.read_csv('../data/olist_order_items_dataset.csv')
reviews = pd.read_csv('../data/olist_order_reviews_dataset.csv')
orders = pd.read_csv('../data/olist_orders_dataset.csv')
sellers = pd.read_csv('../data/olist_sellers_dataset.csv')
products = pd.read_csv('../data/olist_products_dataset.csv')
translation = pd.read_csv('../data/product_category_name_translation.csv', header=0)

## Transformation
After the extraction phase, it’s time to transform the data by identifying relationships between the data points and creating new features that will be useful for future analysis.



The first step will be to gather detailed information about specific datasets and identify the necessary features.

In [3]:
customers.info()
customers = customers.loc[:,["customer_id", "customer_city", "customer_state"]]
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   customer_id     99441 non-null  object
 1   customer_city   99441 non-null  object
 2   customer_state  99441 non-null  object
dtypes: object(3)
memory usage: 2.3+ MB


In [4]:
order_items.info()
order_items = order_items.loc[:,["order_id","product_id","seller_id","price","freight_value"]]
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 5 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   order_id       112650 non-null  object 
 1   product_id     112650 non-null  object 
 2   seller_id      112650 non-null  object 
 3   price          112650 non-nul

In [5]:
reviews.info()
reviews = reviews.loc[:,["order_id","review_score"]]
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   order_id      99224 non-null  object
 1   review_score  99224 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 1.5+ MB


In [6]:
orders.info()
orders = orders.loc[:,["order_id","customer_id","order_delivered_carrier_date","order_delivered_customer_date","order_estimated_delivery_date"]]
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       

In [7]:
sellers.info()
sellers = sellers.loc[:,["seller_id","seller_city","seller_state"]]
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   seller_id     3095 non-null   object
 1   seller_city   3095 non-null   object
 2   seller_state  3095 non-null   object
dtypes: object(3)
memory usage: 72.7+ KB


In [8]:
products.info()
products = products.loc[:,["product_id","product_category_name"]]
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 2 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0 

In [9]:
translation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


After extracting the necessary attributes, we then join the datasets using primary and foreign keys.

In [10]:
#Left Outer Join orders and order_items ON order_id
result = pd.merge(orders, order_items, on="order_id", how="left")

#Left Outer Join result and reviews ON order_id
result = pd.merge(result, reviews, on="order_id", how="left")

#Left Outer Join result and customers ON customer_id
result = pd.merge(result, customers, on="customer_id", how="left")

#Left Outer Join result and sellers ON seller_id
result = pd.merge(result, sellers, on="seller_id", how="left")

#Left Outer Join result and products ON product_id
result = pd.merge(result, products, on="product_id", how="left")

#Left Outer Join result and product_category ON product_category_name
result = pd.merge(result, translation, on="product_category_name", how="left")

result["total_order_value"] = result["price"] + result["freight_value"]
result.rename(columns={"product_category_name_english": "category"}, inplace=True)
result.drop("product_category_name", axis=1, inplace=True)
date_cols = [
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for col in date_cols:
    result[col] = pd.to_datetime(result[col])


After the merge, the final table was reviewed, and the number of NULL values for each column was checked

In [11]:
result.info()
result.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114092 entries, 0 to 114091
Data columns (total 16 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       114092 non-null  object        
 1   customer_id                    114092 non-null  object        
 2   order_delivered_carrier_date   112112 non-null  datetime64[ns]
 3   order_delivered_customer_date  110839 non-null  datetime64[ns]
 4   order_estimated_delivery_date  114092 non-null  datetime64[ns]
 5   product_id                     113314 non-null  object        
 6   seller_id                      113314 non-null  object        
 7   price                          113314 non-null  float64       
 8   freight_value                  113314 non-null  float64       
 9   review_score                   113131 non-null  float64       
 10  customer_city                  114092 non-null  object        
 11  

order_id                            0
customer_id                         0
order_delivered_carrier_date     1980
order_delivered_customer_date    3253
order_estimated_delivery_date       0
product_id                        778
seller_id                         778
price                             778
freight_value                     778
review_score                      961
customer_city                       0
customer_state                      0
seller_city                       778
seller_state                      778
category                         2414
total_order_value                 778
dtype: int64

As you can see, there are quite a few NULL values, particularly in the 'order_delivered_customer_date' column, which contains 3,253 such values.

## Handling Missing Values

Missing values were analyzed and handled based on their meaning:

- Orders without delivery dates were removed, as they were not relevant for delivery analysis
- Rows without price or freight value were removed, as they cannot be used for revenue calculations
- Missing review scores were kept, as not all customers leave reviews

In [12]:
result = result.dropna(subset=["price", "freight_value"])
result = result.dropna(subset=["order_delivered_customer_date", "order_delivered_carrier_date"])

result.info()

<class 'pandas.core.frame.DataFrame'>
Index: 110838 entries, 0 to 114091
Data columns (total 16 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       110838 non-null  object        
 1   customer_id                    110838 non-null  object        
 2   order_delivered_carrier_date   110838 non-null  datetime64[ns]
 3   order_delivered_customer_date  110838 non-null  datetime64[ns]
 4   order_estimated_delivery_date  110838 non-null  datetime64[ns]
 5   product_id                     110838 non-null  object        
 6   seller_id                      110838 non-null  object        
 7   price                          110838 non-null  float64       
 8   freight_value                  110838 non-null  float64       
 9   review_score                   110011 non-null  float64       
 10  customer_city                  110838 non-null  object        
 11  custo

Next, new features were created that will be necessary for future analysis.

In [13]:
#delivery_delay_days
result["delivery_delay_days"] = (result["order_delivered_customer_date"] - result["order_estimated_delivery_date"]).dt.days
result.loc[result["delivery_delay_days"] < 0, "delivery_delay_days"] = 0

#on_time
result["on_time"] = (result["order_estimated_delivery_date"] >= result["order_delivered_customer_date"]).astype(int)

#same_state
result["same_state"] = (result["seller_state"] == result["customer_state"]).astype(int)

#delivery_time
result["delivery_time"] = (result["order_delivered_customer_date"] - result["order_delivered_carrier_date"]).dt.days

The previously created table contains multiple rows for a single order, breaking them down by individual products. Therefore, a separate table for grouped orders has been created. It contains the following columns: ‘total_order_value’, ‘total_freight_price’, ‘items_count’.

In [22]:
result_grouped = result.groupby("order_id")[["price","freight_value","order_id"]].agg({
    "price":"sum",
    "freight_value":"sum",
    "order_id":"count"})
result_grouped.rename(columns={"order_id":"items_count"}, inplace=True)
result_grouped["total_order_value"] = (result_grouped["price"] + result_grouped["freight_value"])

result_grouped.head()


,price,freight_value,items_count,total_order_value
order_id,,,,
00010242fe8c5a6d1ba2dd792cb16214,58.90,13.29,1,72.19
00018f77f2f0320c557190d7a144bdd3,239.90,19.93,1,259.83
000229ec398224ef6ca0657da4fc703e,199.00,17.87,1,216.87
00024acbcdf0a6daa1e931b038114c75,12.99,12.79,1,25.78
00042b26cf59d7ce69dfabb4e55b4fd9,199.90,18.14,1,218.04


## Analysis
The next step will be to conduct an analysis in an effort to find answers to the questions raised.

### Which product categories generate the highest revenue?